README.md
# TradEsp
#### Libreria de de traduccion basica para Arduino.

Esta libreria fue pensada para facilitar el uso del lenguaje Arduino en cursos de programacion orientados a estudiantes no familiarizados con el idioma ingles y los conceptos basicos de programacion.

>*TradEsp fue pensada para un uso exclusivamente pedagogico,* por esa razon la libreria no respeta algunas reglas establecidas por la comunidad Arduino para la creacion de librerias.



# Referencia
#### Funciones Soportadas

Retorna | Funcion | Equivalente
------- | ------- | -----------
|`void`| `escribirDigital()`|  `digitalWrite()` |
|`void`| `espera()`     | `delay()`  
| `void`| `modoPin()` |  `pinMode()` |
|`int`| `leerDigital()`| `digitalRead()`|
| `int`| `leerAnalogico()` | `analogRead()`|
| `void`| `escribirAnalogico()` | `analogWrite()`|
| `void`| `iniciarSerial()`| `Serial.begin()`|
| `void`| `imprimirSerial()`|`Serial.print()`|
| `void`| `tono()`|`tone()`|
| `void`| `noTono()`|`noTone()`|

#### Tipos de Datos y Constantes

##### Tipos de Datos
Tipo | Nombre en TradEsp
------- | -------
`int` | `entero`
`float` | `flotante`
`String` | `cadena`
`boolean` | `booleana`

##### Constantes
Nombre Original | Nombre en TradEsp
------- | -------
`HIGH` | `ENCENDER`
`LOW` | `APAGAR`
`INPUT` | `ENTRADA`
`OUTPUT` | `SALIDA`
`true` | `verdadero`
`false` | `falso`

##### Flujo de Datos
Nombre Original | Nombre en TradEsp
------- | -------
`if` | `Si`
`else` | `SiNo`

# Querido Blink
```Arduino
#include <trad.h>

void setup() {
  //Definimos el pin numero 13 como SALIDA
  modoPin(13, SALIDA);

}

void loop() {
  //Escribimos el valor digital de ENCENDER en el pin numero 13
  escribirDigital(13, ENCENDER);
  //Esperamos 1000 milisegundos
  espera(1000);

  //Apagamos el led 13 y volvemos a esperar.
  escribirDigital(13, APAGAR);
  espera(1000);

}
```


changelog.txt
22/6/2018
*Agregados tipo de dato: char y const
*Agregado soporte para flujo de datos switch, case y break
*Agregado soporte para libreria Keypad.h
21/6/2018
* Separar los metodos privados de los publicos.
* Funciones para el giroscopio comentadas
* Agregados los metodos para la funcion tone()
* Agregadas las constantes para la libreria servo.h


trad.h
//////////////////////////////////////////////////////////////
// Trad_Esp                                                 //
// Libreria de traduccion de funciones basicas de Arduino.  //
// Enero 2017.                                              //
// Marcos Gimenez y Pablo (Malo) Alonso                     //
// Version Base 2.0                                         //
//////////////////////////////////////////////////////////////


#ifndef ARD_ESP_H
#define ARD_ESP_H

#include <Arduino.h>
#include <Wire.h> //SDA: A4 y SCL: A5

///////////////////////////////////
// CONSTANTES DEL NUCLEO ARDUINO//
//////////////////////////////////

//Constantes
#define ENCENDER 1
#define APAGAR 0
#define ENTRADA 0
#define SALIDA 1

//Tipo de Dato
#define entero int
#define flotante float
#define Cadena String
#define booleana boolean
#define constante const
#define caracter char

//Boolean
#define verdadero true
#define falso false

//dataflow
#define Si if
#define SiNo else
#define Mientras while
#define Para for
#define cambio switch
#define caso case
#define salir break

//funciones para servo
#define conectar attach
#define posicion write
#define desconectar detach

//funciones para libreria Keypad.h
#define configTeclas makeKeymap //void, Inicializa el mapa de teclas
#define teclaPresionada getKey //char, Devuelve la tecla presionada si hay alguna
#define esperarTecla waitForKey //Espera hasta que una tecla sea presionada, detiene todo el programa
#define tiempoPresionado setHoldTime //(int) Define el tiempo que se debe presionar para que reconozca la tecla como HOLD
#define tiempoEspera setDebounceTime //(int) Define el tiempo de espera antes de poder presionar otra tecla
#define dispararEvento addEventListener //Agrega un evento a las teclas
#define estado getState //Devuelve un KeyState: IDLE, PRESSED, RELEASED, HOLD
//#define EventoKey KeypadEvent //Tipo de dato
typedef char EventoKey;
#define INACTIVA IDLE //Estado KeyState
#define PRESIONADA PRESSED //Estado KeyState
#define SUELTA RELEASED //Estado KeyState
#define PRESIONANDO HOLD //Estado KeyState


///////////////////////////////////
//CONSTANTES PRIVADAS EXTERNAS//
//////////////////////////////////

//Constantes para Giroscopio
#define MPU9250_ADDRESS 0x68
#define MAG_ADDRESS 0x0C

#define GYRO_FULL_SCALE_250_DPS 0x00
#define GYRO_FULL_SCALE_500_DPS 0x08
#define GYRO_FULL_SCALE_1000_DPS 0x10
#define GYRO_FULL_SCALE_2000_DPS 0x18

#define ACC_FULL_SCALE_2_G 0x00
#define ACC_FULL_SCALE_4_G 0x08
#define ACC_FULL_SCALE_8_G 0x10
#define ACC_FULL_SCALE_16_G 0x18
//Hasta aca giroscopio

///////////////////////////////////
//FUNCIONES DEL NUCLEO DE ARDUINO//
///////////////////////////////////

//digitalRead
int leerDigital(int pin);

//analogRead
int leerAnalogico(int pin);

//analogWrite
void escribirAnalogico(int pin, int valor);

//digitalWrite
void escribirDigital(int pin, int mode);

//pinModo
void modoPin(int pin, int mode);

//Serial.begin
void iniciarSerial(int vel);

//delay
void espera(int tiempo);

// Serial
  template<typename T>
  void imprimirSerial(char *definicion, T valor){
    if (log){
      Serial.print(definicion);
      Serial.print(": ");
      Serial.println(valor);
    }
  }

//tone
void tono(int pin, int freq, int dur);

//noTone
void noTono(int pin);

/////////////////////////////////////////////////////
//FUNCIONES EXTERNAS PARA DISPOSITIVOS DETERMINADOS//
/////////////////////////////////////////////////////

//Sensor distancia ultrasonido
long sensorDistancia(int trig, int echo);

//Giroscopio y acelerometro
int leerAccelGiro(char valor);

//Estas funciones son internas para el uso del giroscopio, no tiene sentido traducirlas (metodos privados?)
void I2Cread(uint8_t Address, uint8_t Register, uint8_t Nbytes, uint8_t* Data);
void I2CwriteByte(uint8_t Address, uint8_t Register, uint8_t Data);
void configAccelGiro();


//7-SEGMENTOS
void print7S(byte addres, String toSend);
void clear7S(byte addres);
void brillo7S(byte addres, byte value);
void coma7S(byte addres, byte decimals);
void testS7S(byte addres);
void convertirD27S(char *tempString, int counter);
//Para el uso de estas funciones se debe conectar el 7S al pin RX (0) del arduino.
void setADDR(byte addrOld, byte addrNew);
void factoryRST();

/*                              LUGARES PANTALLA SPARKFUN
	position	 1	 2	 3	 4	 5	 6	 7	 8	 9	10	11	12	13	14	15	16
	line 1		128	129	130	131	132	133	134	135	136	137	138	139	140	141	142	143
	line 2		192	193	194	195	196	197	198	199	200	201	202	203	204	205	206	207
*/
#endif


privados.cpp
//////////////////////////////////////////////////////////////
// Trad_Esp                                                 //
// Libreria de traduccion de funciones basicas de Arduino.  //
// Enero 2017.                                              //
// Marcos Gimenez y Pablo (Malo) Alonso                     //
// Version Base 2.0                                         //
//////////////////////////////////////////////////////////////


#include "trad.h"

/////////////////////////////////////////////////////////////
// METODOS PRIVADOS PARA EL FUNCIONAMIENTO DEL GIROSCOPIO //
////////////////////////////////////////////////////////////

void I2Cread(uint8_t Address, uint8_t Register, uint8_t Nbytes, uint8_t* Data)
{
	// Set register address
	Wire.beginTransmission(Address);
	Wire.write(Register);
	Wire.endTransmission();

	// Read Nbytes
	Wire.requestFrom(Address, Nbytes);
	uint8_t index=0;
	while (Wire.available())
	Data[index++]=Wire.read();
}

void I2CwriteByte(uint8_t Address, uint8_t Register, uint8_t Data)
{
	// Set register address
	Wire.beginTransmission(Address);
	Wire.write(Register);
	Wire.write(Data);
	Wire.endTransmission();
}

void configAccelGiro()
{

	// CONFIGURACION DE LAS CONSTANTES Y FUNCIONES//
	Wire.begin();
	// Set accelerometers low pass filter at 5Hz
	I2CwriteByte(MPU9250_ADDRESS,29,0x06);
	// Set gyroscope low pass filter at 5Hz
	I2CwriteByte(MPU9250_ADDRESS,26,0x06);
	// Configure gyroscope range
	I2CwriteByte(MPU9250_ADDRESS,27,GYRO_FULL_SCALE_1000_DPS);
	// Configure accelerometers range
	I2CwriteByte(MPU9250_ADDRESS,28,ACC_FULL_SCALE_4_G);
	// Set by pass mode for the magnetometers
	I2CwriteByte(MPU9250_ADDRESS,0x37,0x02);

	 /*
	// Request continuous magnetometer measurements in 16 bits
	 I2CwriteByte(MAG_ADDRESS,0x0A,0x16);
	 pinMode(13, OUTPUT);
	 Timer1.initialize(10000); // initialize timer1, and set a 1/2 second period
	 Timer1.attachInterrupt(callback); // attaches callback() as a timer overflow interrupt
  */
}


trad.cpp
//////////////////////////////////////////////////////////////
// Trad_Esp                                                 //
// Libreria de traduccion de funciones basicas de Arduino.  //
// Enero 2017.                                              //
// Marcos Gimenez y Pablo (Malo) Alonso                     //
// Version Base 2.0                                         //
//////////////////////////////////////////////////////////////


#include "trad.h"

////////////////////////////////////////////
// METODOS Y FUNCIONES DEL NUCLEO ARDUINO //
////////////////////////////////////////////


//digitalWrite
void escribirDigital(int pin, int mode)
{
    digitalWrite(pin, mode);

}

//delay
void espera(int tiempo)
{
    delay(tiempo);
}

//pinMode
void modoPin(int pin, int mode)
{
    if (mode == 1) //1  = OUTPUT
    {
        mode == OUTPUT;
    } else if (mode == 0) // 0 = INPUT
    {
        mode == INPUT;
    }

    pinMode(pin, mode);
}

//digitalRead
int leerDigital(int pin)
{
    int val = digitalRead(pin);
    return val;
}

//analogRead
int leerAnalogico(int pin)
{
    int val = analogRead(pin);
    return val;
}

//analogWrite
void escribirAnalogico(int pin, int valor)
{
    analogWrite(pin, valor);
}

//Serial.begin
void iniciarSerial(int vel)
{
    Serial.begin(vel);
}

//tone
void tono(int pin, int freq, int dur){
    tone(pin, freq, dur);
}

//noTone
void noTone(int pin){
    noTone(pin);
}

//////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////
// METODOS AGREGADOS PARA SHIELDS Y COMPONENTES ESPECIFICOS//
/////////////////////////////////////////////////////////////
/////////////////////////////////////////////////////////////


//---------------------------------------------------------//


// METODO PARA LEER LOS VALORES DEL ACELEROMETRO Y GIROSCOPIO - METODO PUBLICO
int leerAccelGiro(char valor)
{
	// ::: accelerometer and gyroscope :::

	// Read accelerometer and gyroscope
	uint8_t Buf[14];
	I2Cread(MPU9250_ADDRESS,0x3B,14,Buf);

	// Create 16 bits values from 8 bits data
	/*
		// Accelerometer
		int16_t ax=-(Buf[0]<<8 | Buf[1]);
		int16_t ay=-(Buf[2]<<8 | Buf[3]);
		int16_t az=Buf[4]<<8 | Buf[5];

		// Gyroscope
		int16_t gx=-(Buf[8]<<8 | Buf[9]);
		int16_t gy=-(Buf[10]<<8 | Buf[11]);
		int16_t gz=Buf[12]<<8 | Buf[13];
	*/

	// Accelerometer
	if (valor ==1)
	{
		return -(Buf[0]<<8 | Buf[1]);
	}
	if (valor ==2)
	{
		return -(Buf[2]<<8 | Buf[3]);
	}
	if (valor ==3)
	{
		return Buf[4]<<8 | Buf[5];
	}

	// Gyroscope
	if (valor ==4)
	{
		return -(Buf[8]<<8 | Buf[9]);
	}
	if (valor ==5)
	{
		return -(Buf[10]<<8 | Buf[11]);
	}
	if (valor ==6)
	{
		return Buf[12]<<8 | Buf[13];
	}

}



//{///////////////////////////////////////7 SEGMENTOS////////////////////////////////////////////////
void print7S(byte addres, String toSend)
{
  Wire.beginTransmission(addres);
  for (int i = 0; i < 4; i++)
  {
    Wire.write(toSend[i]);
  }
  Wire.endTransmission();
}

void clear7S(byte addres)
{
  Wire.beginTransmission(addres);
  Wire.write(0x76);  // Clear display command
  Wire.endTransmission();
}

// Set the displays brightness. Should receive byte with the value
//  to set the brightness to
//  dimmest------------->brightest
//     0--------127--------255
void brillo7S(byte addres, byte value)
{
  Wire.beginTransmission(addres);
  Wire.write(0x7A);  // Set brightness command byte
  Wire.write(value);  // brightness data byte
  Wire.endTransmission();
}

// Turn on any, none, or all of the decimals.
//  The six lowest bits in the decimals parameter sets a decimal
//  (or colon, or apostrophe) on or off. A 1 indicates on, 0 off.
//  [MSB] (X)(X)(Apos)(Colon)(Digit 4)(Digit 3)(Digit2)(Digit1)
void coma7S(byte addres, byte decimals)
{
  Wire.beginTransmission(addres);
  Wire.write(0x77);
  Wire.write(decimals);
  Wire.endTransmission();
}

void testS7S(byte addres)
{
  Wire.beginTransmission(addres);
  Wire.write(0x76);  // Clear the display
  delay(1000);
  Wire.write('t');
  Wire.write('e');
  Wire.write('s');
  Wire.write('t');
  Wire.endTransmission();
}

void convertirD27S(char *tempString, int counter)
{
	sprintf(tempString, "%4d", counter);
}


void setADDR(byte addrOld, byte addrNew)////////ESTO ES PARA CONFIGURAR EL DISPLAY PERO SE DEBE HACER CON HARDWARE SERIAL
{
  int baudRates[12] = {2400, 4800, 9600, 14400, 19200, 38400,
                       57600, 76800, 115200, 250000, 500000, 1000000
                      };
  for (int i = 0; i < 12; i++)
  {
    Serial.begin(baudRates[i]);  // Set new baud rate
    delay(10);  // Arduino needs a moment to setup serial
    Serial.write(0x80);  // I2C Address Config command
    Serial.write(addrNew);  // Set 7-bit address to 0x42
  }
  addrOld = addrNew;
  Serial.begin(9600);
  delay(10);  // Arduino needs a moment to setup serial
  Serial.write(0x76);  // Clear the display
  delay(10);
  Serial.write('t');
  delay(10);
  Serial.write('e');
  delay(10);
  Serial.write('s');
  delay(10);
  Serial.write('t');
}
/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

void factoryRST()////////ESTO ES PARA CONFIGURAR EL DISPLAY PERO SE DEBE HACER CON HARDWARE SERIAL
{
  int baudRates[12] = {2400, 4800, 9600, 14400, 19200, 38400,
                       57600, 76800, 115200, 250000, 500000, 1000000
                      };
  for (int i = 0; i < 12; i++)
  {
    Serial.begin(baudRates[i]);  // Set new baud rate
    delay(10);  // Arduino needs a moment to setup serial
    Serial.write(0x81);  // Send factory reset command
  }

  Serial.begin(9600);
  delay(10);  // Arduino needs a moment to setup serial
  Serial.write(0x76);  // Clear the display
  Serial.write('t');
  delay(10);
  Serial.write('e');
  delay(10);
  Serial.write('s');
  delay(10);
  Serial.write('t');
}

//////////////////////////////////////////////////////////////////////////////////////////////////}

//{/////////////////////////////////SENSOR DISTANCIA ULTRASONIDO/////////////////////////////////////
long sensorDistancia(int trig,int echo)
{
	long distancia;
	long tiempo;

	digitalWrite(trig,LOW); // Por cuestión de estabilización del sensor
	delayMicroseconds(5);
	digitalWrite(trig, HIGH); // envío del pulso ultrasónico
	delayMicroseconds(10);

	// Función para medir la longitud del pulso entrante. Mide el tiempo que transcurrido entre el envío
	// del pulso ultrasónico y cuando el sensor recibe el rebote, es decir: desde que el pin 12 empieza
	// a recibir el rebote, HIGH, hasta que deja de hacerlo, LOW, la longitud del pulso entrante
	tiempo=pulseIn(echo, HIGH);

	distancia= int(0.017*tiempo); /*fórmula para calcular la distancia obteniendo un valor entero*/

	return distancia;
}
////////////////////////////////////////FIN SENSOR DISTANCIA ULTRASONIDO//////////////////////////////////}
